File: HisabCore/Sources/HisabCore/Analytics.swift
import Foundation

/// One "counted" row for analytics. The app layer feeds every payment-app transaction
/// plus unmatched bank rows (category "Miscellaneous" when no rule hits); matched bank
/// rows (reconciliation evidence) and cross-bank self transfers are excluded upstream,
/// so nothing here is double-counted.
public struct AnalyticsTxn: Sendable {
    public let month: YearMonth
    public let amountPaise: Int64
    public let direction: Direction
    public let category: String
    public let merchant: String
    public let sourceKind: SourceKind

    public init(month: YearMonth, amountPaise: Int64, direction: Direction,
                category: String, merchant: String, sourceKind: SourceKind) {
        self.month = month
        self.amountPaise = amountPaise
        self.direction = direction
        self.category = category
        self.merchant = merchant
        self.sourceKind = sourceKind
    }
}

public struct MonthStats: Sendable, Equatable {
    public let month: YearMonth
    public let spendPaise: Int64
    public let incomePaise: Int64

    public var netPaise: Int64 { incomePaise - spendPaise }
}

public enum Analytics {
    public static func monthStats(_ txns: [AnalyticsTxn], month: YearMonth) -> MonthStats {
        let monthTxns = txns.filter { $0.month == month }
        let spend = monthTxns.filter { $0.direction == .debit }.reduce(Int64(0)) { $0 + $1.amountPaise }
        let income = monthTxns.filter { $0.direction == .credit }.reduce(Int64(0)) { $0 + $1.amountPaise }
        return MonthStats(month: month, spendPaise: spend, incomePaise: income)
    }

    public static func trend(_ txns: [AnalyticsTxn], endingAt end: YearMonth, count: Int) -> [MonthStats] {
        YearMonth.months(from: end.advanced(by: -(count - 1)), through: end)
            .map { monthStats(txns, month: $0) }
    }

    /// Sum of debit amounts per grouping key, largest first (ties alphabetical).
    /// Kept as explicit steps — older Swift compilers time out on the fused chain.
    private static func debitTotals(_ txns: [AnalyticsTxn], month: YearMonth,
                                    by key: (AnalyticsTxn) -> String) -> [(String, Int64)] {
        let debits = txns.filter { $0.month == month && $0.direction == .debit }
        var totals: [String: Int64] = [:]
        for txn in debits {
            totals[key(txn), default: 0] += txn.amountPaise
        }
        return totals.sorted { lhs, rhs in
            if lhs.value == rhs.value { return lhs.key < rhs.key }
            return lhs.value > rhs.value
        }
    }

    public static func categoryBreakdown(_ txns: [AnalyticsTxn], month: YearMonth,
                                         top: Int) -> [(category: String, paise: Int64)] {
        let totals = debitTotals(txns, month: month, by: \.category)
        var result = totals.map { (category: $0.0, paise: $0.1) }
        guard result.count > top else { return result }
        var rest: Int64 = 0
        for entry in result.dropFirst(top) {
            rest += entry.paise
        }
        result = Array(result.prefix(top))
        result.append((category: "Other", paise: rest))
        return result
    }

    public static func topMerchants(_ txns: [AnalyticsTxn], month: YearMonth,
                                    top: Int) -> [(merchant: String, paise: Int64)] {
        debitTotals(txns, month: month, by: \.merchant).prefix(top)
            .map { (merchant: $0.0, paise: $0.1) }
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/analytics.dart
/// Analytics over counted rows. Port of Analytics.swift.
library;

import 'domain.dart';
import 'year_month.dart';

/// One "counted" row: payment-app transactions plus unmatched bank rows;
/// matched bank evidence and self transfers are excluded upstream.
class AnalyticsTxn {
  final YearMonth month;
  final int amountPaise;
  final Direction direction;
  final String category;
  final String merchant;
  final SourceKind sourceKind;
  const AnalyticsTxn({
    required this.month,
    required this.amountPaise,
    required this.direction,
    required this.category,
    required this.merchant,
    required this.sourceKind,
  });
}

class MonthStats {
  final YearMonth month;
  final int spendPaise;
  final int incomePaise;
  const MonthStats(
      {required this.month, required this.spendPaise, required this.incomePaise});
  int get netPaise => incomePaise - spendPaise;

  @override
  bool operator ==(Object other) =>
      other is MonthStats &&
      other.month == month &&
      other.spendPaise == spendPaise &&
      other.incomePaise == incomePaise;
  @override
  int get hashCode => Object.hash(month, spendPaise, incomePaise);
}

class CategorySlice {
  final String category;
  final int paise;
  const CategorySlice(this.category, this.paise);
}

class MerchantSlice {
  final String merchant;
  final int paise;
  const MerchantSlice(this.merchant, this.paise);
}

class Analytics {
  static MonthStats monthStats(List<AnalyticsTxn> txns, YearMonth month) {
    var spend = 0;
    var income = 0;
    for (final txn in txns) {
      if (txn.month != month) continue;
      if (txn.direction == Direction.debit) {
        spend += txn.amountPaise;
      } else {
        income += txn.amountPaise;
      }
    }
    return MonthStats(month: month, spendPaise: spend, incomePaise: income);
  }

  static List<MonthStats> trend(List<AnalyticsTxn> txns,
      {required YearMonth endingAt, required int count}) {
    return YearMonth.monthsFromThrough(endingAt.advancedBy(-(count - 1)), endingAt)
        .map((m) => monthStats(txns, m))
        .toList();
  }

  static List<MapEntry<String, int>> _debitTotals(
      List<AnalyticsTxn> txns, YearMonth month, String Function(AnalyticsTxn) key) {
    final totals = <String, int>{};
    for (final txn in txns) {
      if (txn.month != month || txn.direction != Direction.debit) continue;
      totals[key(txn)] = (totals[key(txn)] ?? 0) + txn.amountPaise;
    }
    final entries = totals.entries.toList()
      ..sort((l, r) =>
          l.value == r.value ? l.key.compareTo(r.key) : r.value.compareTo(l.value));
    return entries;
  }

  static List<CategorySlice> categoryBreakdown(
      List<AnalyticsTxn> txns, YearMonth month, {required int top}) {
    final totals = _debitTotals(txns, month, (t) => t.category)
        .map((e) => CategorySlice(e.key, e.value))
        .toList();
    if (totals.length <= top) return totals;
    var rest = 0;
    for (final slice in totals.skip(top)) {
      rest += slice.paise;
    }
    return [...totals.take(top), CategorySlice('Other', rest)];
  }

  static List<MerchantSlice> topMerchants(
      List<AnalyticsTxn> txns, YearMonth month, {required int top}) {
    return _debitTotals(txns, month, (t) => t.merchant)
        .take(top)
        .map((e) => MerchantSlice(e.key, e.value))
        .toList();
  }
}
